clippy_lints/src/methods/iter_filter.rs

```diff
@@ -7,81 +9,116 @@ use clippy_utils::source::{indent_of, reindent_multiline};
 use clippy_utils::{is_trait_method, peel_blocks, span_contains_comment};
 use rustc_errors::Applicability;
 use rustc_hir as hir;
-use rustc_hir::def::Res;
 use rustc_hir::QPath;
 use rustc_span::symbol::{sym, Symbol};
 use rustc_span::Span;
 use std::borrow::Cow;
 
-fn is_method(cx: &LateContext<'_>, expr: &hir::Expr<'_>, method_name: Symbol) -> bool {
-    match &expr.kind {
-        hir::ExprKind::Path(QPath::TypeRelative(_, mname)) => mname.ident.name == method_name,
-        hir::ExprKind::Path(QPath::Resolved(_, segments)) => {
-            segments.segments.last().unwrap().ident.name == method_name
+fn is_method(cx: &LateContext<'_>, expr: &hir::Expr<'_>, method_name: Symbol, expected_symbol: Symbol) -> bool {
+    let did = cx.tcx.get_diagnostic_item(expected_symbol).unwrap();
+    let expected_ty = cx.tcx.type_of(did).skip_binder();
+
+    match expr.kind {
+        hir::ExprKind::Path(QPath::TypeRelative(ty, mname)) => {
+            // from hir::ty to rustc_middle::ty
+            let ty = cx.typeck_results().node_type(ty.hir_id);
+            expected_ty.ty_adt_def() == ty.ty_adt_def() && mname.ident.name == method_name
         },
-        hir::ExprKind::MethodCall(segment, _, _, _) => segment.ident.name == method_name,
         hir::ExprKind::Closure(&hir::Closure { body, .. }) => {
-            let body = cx.tcx.hir().body(body);
+            let body: &hir::Body<'_> = cx.tcx.hir().body(body);
             let closure_expr = peel_blocks(body.value);
-            let arg_id = body.params[0].pat.hir_id;
-            match closure_expr.kind {
-                hir::ExprKind::MethodCall(hir::PathSegment { ident, .. }, receiver, ..) => {
-                    if ident.name == method_name
-                        && let hir::ExprKind::Path(path) = &receiver.kind
-                        && let Res::Local(ref local) = cx.qpath_res(path, receiver.hir_id)
-                    {
-                        return arg_id == *local;
-                    }
-                    false
-                },
-                _ => false,
+            if let hir::ExprKind::MethodCall(hir::PathSegment { ident, .. }, receiver, ..) = closure_expr.kind
+                && ident.name == method_name
+                && let hir::ExprKind::Path(path) = receiver.kind
+                && let QPath::Resolved(None, path) = path
+                && let Path {
+                    res: hir::def::Res::Local(_),
+                    ..
+                } = path
+            {
+                return true;
             }
+            false
```

Comment:
The handling of the closure feels a bit hacky. The essence is that it checks for a single expression after peeling that is of the form `a.is_ok`.